pkg/package-server/provider/registry.go

```diff
@@ -0,0 +1,472 @@
+	logger := logrus.WithFields(logrus.Fields{
+		"action":    "List PackageManifests",
+		"namespace": namespace,
+	})
+
+	p.mu.RLock()
+	defer p.mu.RUnlock()
+
+	pkgs := []v1alpha1.PackageManifest{}
+	for _, client := range p.clients {
+		if client.source.GetNamespace() == namespace || client.source.GetNamespace() == p.globalNamespace || namespace == "" {
+			logger.Debugf("found CatalogSource %s", client.source.GetName())
+
+			stream, err := client.ListPackages(context.Background(), &api.ListPackageRequest{})
+			if err != nil {
+				logger.Warnf("error getting stream: %v", err)
+				return nil, err
+			}
+			for {
+				logger.Info("in for loop getting data from list stream...")
+				pkgName, err := stream.Recv()
+				if err == io.EOF {
+					logger.Info("end of stream")
+					break
+				}
+
+				if err != nil {
+					logger.Warnf("some error while getting data: %v", err)
+					return nil, err
+				}
+				pkg, err := client.GetPackage(context.Background(), &api.GetPackageRequest{Name: pkgName.GetName()})
+				if err != nil {
+					logger.Warnf("some error while getting package: %v", err)
+					return nil, err
+				}
+				newPkg, err := toPackageManifest(pkg, client)
+				pkgs = append(pkgs, *newPkg)
+			}
+		}
+	}
+
+	return &v1alpha1.PackageManifestList{Items: pkgs}, nil
+}
+
+func (p *RegistryProvider) forward(ctx context.Context, logger *logrus.Entry, out chan v1alpha1.PackageManifest, transform forwarder.Transform) {
+	from := make(chan interface{})
+	to := make(chan interface{})
+
+	go func() {
+		for {
+			select {
+			case event := <-p.pkgTopic:
+				from <- event
+			case event := <-to:
+				if e, ok := event.(*pkgEvent); ok {
+					out <- *(e.pkg)
+				} else {
+					logger.Warn("could not assert package event type")
+				}
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+
+	go func() {
+		err := p.forwarder.Forward(ctx, from, to, transform)
+		if err != nil {
+			logger.WithField("err", err.Error()).Warn("error forwarding events")
+		}
+	}()
+}
+
+func pkgTransform(t pkgEventType) forwarder.Transform {
+	return func(obj interface{}) (interface{}, bool) {
+		if e, ok := obj.(*pkgEvent); ok && e.eventType == t {
+			return e.pkg, true
+		}
+		return nil, false
+	}
+}
+
+func (p *RegistryProvider) Subscribe(ctx context.Context, namespace string) (PackageChan, PackageChan, PackageChan, error) {
+	logger := logrus.WithFields(logrus.Fields{
+		"Action":    "PackageManifest Subscribe",
+		"namespace": namespace,
+	})
+
+	add := make(chan v1alpha1.PackageManifest)
+	modify := make(chan v1alpha1.PackageManifest)
+	delete := make(chan v1alpha1.PackageManifest)
+
+	p.mu.RLock()
+	defer p.mu.RUnlock()
+
+	for _, client := range p.clients {
+		if client.source.GetNamespace() == namespace || client.source.GetNamespace() == p.globalNamespace || namespace == "" {
+			logger.Debugf("found CatalogSource %s", client.source.GetName())
+			// Start up forwarders.
+			p.forward(ctx, logger, add, pkgTransform(addPkg))
```

Comment:
super nit: `pkgTransform` to me suggests that this `Transform` will mutate the events. I'd prefer a naming that suggested that it's just a filter, e.g. `addEventFilter := pkgTransform(addPkg)`